views/js/runner/provider/qti.js

```diff
@@ -190,118 +208,103 @@ define([
-                var performSubmit = function performSubmit(){
-                    //we submit the responses
-                    var submitItem = self.getProxy().submitItem(context.itemUri, itemRunner.getState(), itemRunner.getResponses(), params)
-                        .then(function(result){
-                            return new Promise(function(resolve, reject){
-
-                                if (result.notAllowed) {
-                                    // the context might be updated
-                                    if (result.testContext) {
-                                        self.setTestContext(result.testContext);
-                                        context = self.getTestContext();
-                                    }
 
-                                    if (result.message) {
-                                        self.trigger('alert.notallowed',
-                                            result.message,
-                                            function() {
-                                                self.trigger('resumeitem');
-                                            }
-                                        );
-                                    }
+                        self.getProxy()
+                            .callItemAction(context.itemUri, action, params)
+                            .then(function(results){
 
-                                    return reject(true);
+                                if(results.testContext){
+                                    self.setTestContext(results.testContext);
                                 }
 
-                                if (result.itemSession) {
-                                    context.itemAnswered = result.itemSession.itemAnswered;
+                                if (results.testMap) {
+                                    self.setTestMap(results.testMap);
                                 }
 
-                                if(result.displayFeedbacks === true && result.feedbacks && result.itemSession){
+                                updateStats();
 
-                                    itemRunner.renderFeedbacks(result.feedbacks, result.itemSession, function(queue){
-                                        self.trigger('modalFeedbacks', queue, resolve);
-                                    });
+                                if (results.itemSession) {
+                                    context.itemAnswered = results.itemSession.itemAnswered;
 
+                                    if(results.displayFeedbacks === true && results.feedbacks) {
+                                        self.itemRunner.renderFeedbacks(results.feedbacks, results.itemSession, function(queue){
+                                            self.trigger('modalFeedbacks', queue, function(){
+                                                computeNext(action, _.omit(params, 'itemState', 'itemResponse'));
+                                            });
```

Comment:
Maybe I am wrong, but is there chance the feedbacks be displayed twice? There is two places where the feedbacks are handled. I suppose they are sent by the server only after a response submit?